File: 2021/Day10/Day10.scala
import scala.io.Source
import scala.annotation.tailrec

@main def day10 =
    def isClose = Set(')', ']', '}', '>').contains

    def score(symbol: Char) =
        Map((')', 3), (']', 57), ('}', 1197), ('>', 25137)).get(symbol).get

    def cost(symbol: Char) =
        Map((')', 1L), (']', 2L), ('}', 3L), ('>', 4L)).get(symbol).get

    def findClose(symbol: Char) =
        Map(('(', ')'), ('[', ']'), ('{', '}'), ('<', '>')).get(symbol).get

    def analyze(state: (List[Char], Boolean), current: Char) =
        val (symbols, corrupted) = state
        if corrupted then state
        else
            symbols match
                case Nil =>
                    if isClose(current) then (current :: Nil, true)
                    else (current :: Nil, false)
                case s @ symbol :: symbols =>
                    current match
                        case ')' =>
                            if symbol == '(' then (symbols, false)
                            else (current :: s, true)
                        case ']' =>
                            if symbol == '[' then (symbols, false)
                            else (current :: s, true)
                        case '}' =>
                            if symbol == '{' then (symbols, false)
                            else (current :: s, true)
                        case '>' =>
                            if symbol == '<' then (symbols, false)
                            else (current :: s, true)
                        case _ => (current :: s, false)

    def repair(symbols: List[Char]) =
        @tailrec
        def loop(instructions: List[Char], symbols: List[Char]): List[Char] =
            symbols match
                case Nil => instructions
                case symbol :: symbols =>
                    loop((findClose(symbol) :: instructions), symbols)

        loop(Nil, symbols).reverse

    def firstPart(symbols: Vector[Array[Char]]) =
        symbols
            .map(_.foldLeft((Nil, false))(analyze))
            .filter(_._2)
            .map(r => score(r._1.head))
            .sum

    def secondPart(symbols: Vector[Array[Char]]) =
        val costs = symbols
            .map(_.foldLeft((Nil, false))(analyze))
            .filter(!_._2)
            .map(r =>
                repair(r._1)
                    .map(cost)
                    .foldLeft(0L)((total, cost) => total * 5L + cost)
            )
            .sortBy(identity)

        costs(costs.size / 2)

    val input =
        Source.fromFile("input.txt").getLines.map(_.toCharArray).toVector

    println(s"First part: ${firstPart(input)}")
    println(s"Second part: ${secondPart(input)}")

File: 2021/archived/Day10/Day10.fs
module Day10

open System.IO

let corrupted = snd

let isClose symbol = List.contains symbol [ ')'; ']'; '}'; '>' ]

let score symbol =
    match symbol with
    | ')' -> 3
    | ']' -> 57
    | '}' -> 1197
    | '>' -> 25137
    | _ -> failwithf "Bang bang!! %c" symbol

let cost symbol =
    match symbol with
    | ')' -> 1UL
    | ']' -> 2UL
    | '}' -> 3UL
    | '>' -> 4UL
    | _ -> failwithf "Bang bang!! %c" symbol

let findClose symbol =
    match symbol with
    | '(' -> ')'
    | '{' -> '}'
    | '[' -> ']'
    | '<' -> '>'
    | _ -> failwithf "Bang bang!! %c" symbol

let analyze (state as (symbols, corrupted)) current  =
    if corrupted then state
    else
        match symbols with
        | [] -> if isClose current then ([current], true)
                else ([current], false)
        | s as (symbol :: symbols) ->
            match current with
            | ')' -> if symbol = '(' then (symbols, false) else (current :: s, true)
            | ']' -> if symbol = '[' then (symbols, false) else (current :: s, true)
            | '}' -> if symbol = '{' then (symbols, false) else (current :: s, true)
            | '>' -> if symbol = '<' then (symbols, false) else (current :: s, true)
            | _ -> (current :: s, false)

let repair symbols =
    let rec loop instructions symbols =
        match symbols with
        | [] -> instructions
        | symbol :: symbols ->
            loop ((symbol |> findClose) :: instructions) symbols

    loop [] symbols |> List.rev

let firstPart input =
    input
    |> Array.map (Array.fold analyze ([], false))
    |> Array.filter corrupted
    |> Array.map (fst >> List.head >> score)
    |> Array.sum

let secondPart input =
    let costs = input
                |> Array.map (Array.fold analyze ([], false))
                |> Array.filter (not << corrupted)
                |> Array.map (fst >> repair)
                |> Array.map (List.map cost >> List.fold (fun total cost -> total * 5UL + cost) 0UL)
                |> Array.sort

    costs.[costs.Length / 2]

let input = File.ReadAllLines "input.txt" |> Array.map Seq.toArray

input |> firstPart |> printfn "First part: %d"

input |> secondPart |> printfn "Second part: %d"
